src/index.ts

```diff
@@ -0,0 +1,313 @@
+    return this._mode;
+  }
+
+  /**
+   * Dispose of this object.
+   */
+  dispose(): void {
+    this._disposed = true;
+    clearSignalData(this);
+    this._model = null;
+    this._view = null;
+  }
+
+  /**
+   * Get the disposed status of this object.
+   */
+  get isDisposed(): boolean {
+    return this._disposed;
+  }
+
+  /**
+   * Flag that determines whether to show line numbers.
+   */
+  showLineNumbers = true;
+
+  /**
+   * Flag that determines whether the view is read-only.
+   */
+  readOnly = true;
+
+  /**
+   * Number of spaces to use for a tab.
+   */
+  tabSize = 2;
+
+  private _updateBuffer(): void {
+    if (!this.isDisposed) {
+      var data = this._view.getContents();
+      this._buffer = data;
+    }
+  }
+
+  private _startBufferTimer(): void {
+    clearTimeout(this._bufferTimeoutId);
+    this._bufferTimeoutId = setTimeout(() => {
+      this._updateBuffer();
+    }, this._bufferDelay);
+  }
+
+  private _mode = '';
+  private _disposed = false;
+  private _view: IEditorWidget = null;
+  private _model: IEditorModel;
+  private _buffer = '';
+  /**
+   * The delay in milliseconds between view buffer-fetches.
+   */
+  private _bufferDelay = 1000;
+  /**
+   * The id returned by setTimeout.
+   */
+  private _bufferTimeoutId = 0;
+}
+
+/**
+ * The interface to fulfil in order to implement
+ * an EditorWidget.
+ */
+export
+interface IEditorWidget {
+  /**
+   * Get the entire contents of the current editor.
+   */
+  getContents(): string;
+}
+
+/**
+ * A widget used for entering text/code.
+ *
+ * #### Notes
+ * This class happens to use Codemirror as the hosted widget for
+ * text entry, but Codemirror is entirely encapsulated
+ * within this class. As long as the IEditorWidget interface is
+ * adhered to, any other text editor widget can be easily
+ * swapped in.
+ */
+export
+class EditorWidget extends Widget implements IEditorWidget {
+
+  /**
+   * A signal which is emitted when the view contents change.
+   *
+   * #### Notes
+   * This is designed for the view-model. The data is *not* passed
+   * as the signal argument in order that the view-model can decide
+   * when to pull the potentially large data string from the view.
+   */
+  static contentsChangedSignal = new Signal<IEditorWidget, boolean>();
+
+  /**
+   * A pure delegate for [[contentsChangedSignal]].
+   */
+  static getContentsChanged(item: IEditorWidget): ISignal<IEditorWidget, boolean> {
+    return EditorWidget.contentsChangedSignal.bind(item);
+  }
+
+  /**
+   * Construct an EditorWidget.
+   */
+  constructor(vm: IEditorViewModel) {
+    super();
+    this.addClass('EditorWidget');
+
+    var config = {
+      mode: vm.mode(),
+      lineNumbers: vm.showLineNumbers,
+      tabSize: vm.tabSize
+    };
+    this._editor = CodeMirror(this.node, config);
+    this._editor.on("change", () => {
+      EditorWidget.getContentsChanged(this).emit(true);
+    });
+
+    this._viewmodel = vm;
+    this._viewmodel.setView(this);
```

Comment:
The view model should not know about the view. Instead, the view should connect to signals on the view model.